cmd/dashboard/upgrade.go

```diff
@@ -106,75 +143,148 @@ type BatchUpgrade struct {
-	podsStatus map[string]config.UpgradeStatus
+	lock            sync.Mutex
+	podsStatus      map[string]config.UpgradeStatus
+	status          config.UpgradeStatus
+	crtBatchStatus  config.UpgradeStatus
+	nextBatchStatus config.UpgradeStatus
+	crtBatch        int
 }
 
 func (u *BatchUpgrade) Run(ctx context.Context) {
-	if u.conf.Parallel > 50 {
-		fmt.Printf("%s BATCH-FAIL parallel should not exceed 50\n", time.Now().Format(time.DateTime))
+	if len(u.conf.Batches) == 0 {
+		logger("BATCH-SUCCESS no batch found")
+		u.status = config.Success
+		u.flushStatus(ctx)
 		return
 	}
-	for _, batch := range u.conf.Batches {
-		var (
-			limiter  = make(chan struct{}, u.conf.Parallel)
-			resultCh = make(chan error)
-			wg       sync.WaitGroup
-		)
-		go func() {
-			defer func() {
-				wg.Wait()
-				close(resultCh)
-				close(limiter)
-			}()
-			for _, mp := range batch {
-				mp2 := mp
-				select {
-				case <-ctx.Done():
-					return
-				case limiter <- struct{}{}:
-				}
+	if u.conf.Parallel > 50 {
+		logger("BATCH-FAIL parallel should not exceed 50")
+		u.panic(ctx)
+	}
 
-				wg.Add(1)
-				go func() {
-					defer func() {
-						wg.Done()
-						<-limiter
-					}()
-					resultCh <- u.triggerUpgrade(ctx, &mp2)
-				}()
-			}
-		}()
+	handleFinalStatus := func() {
+		if u.status == config.Fail {
+			logger("BATCH-FAIL some pods upgrade failed")
+			u.panic(ctx)
+		}
+		if u.status == config.Success {
+			logger("BATCH-SUCCESS all pods upgraded successfully")
+		}
+		u.flushStatus(ctx)
+	}
+
+	t := time.NewTicker(1 * time.Second)
+	defer t.Stop()
 
-		for oneErr := range resultCh {
-			if oneErr != nil {
-				if !u.conf.IgnoreError {
-					fmt.Printf("%s BATCH-FAIL some pods upgrade failed\n", time.Now().Format(time.DateTime))
-					return
+	for {
+		select {
+		case <-ctx.Done():
+			logger("BATCH-FAIL upgrade timeout")
+			u.panic(ctx)
+			return
+		case <-t.C:
+			if u.crtBatchStatus == config.Fail && !u.conf.IgnoreError {
+				u.status = config.Fail
+				handleFinalStatus()
+				return
+			}
+			if u.crtBatch > len(u.conf.Batches) {
+				u.status = u.crtBatchStatus
+				handleFinalStatus()
+				return
+			}
+			switch u.nextBatchStatus {
+			case config.Pending:
+				if u.crtBatchStatus == config.Pending || u.crtBatchStatus == config.Success || (u.crtBatchStatus == config.Fail && u.conf.IgnoreError) {
+					u.crtBatch++
+					go u.processBatch(ctx)
 				}
+			case config.Pause:
+			case config.Stop:
+				u.status = config.Stop
+				handleFinalStatus()
+				return
 			}
 		}
+	}
+}
 
-		if errors.Is(ctx.Err(), context.DeadlineExceeded) {
-			log.Error(ctx.Err(), "upgrade timeout")
-			fmt.Printf("%s BATCH-FAIL upgrade timeout\n", time.Now().Format(time.DateTime))
+func (u *BatchUpgrade) processBatch(ctx context.Context) {
+	if u.crtBatch > len(u.conf.Batches) {
+		return
+	}
+	u.setCrtBatchStatus(config.Running)
+	var (
+		wg       sync.WaitGroup
+		resultCh = make(chan error, u.conf.Parallel)
+		batch    = u.conf.Batches[u.crtBatch-1]
```

Comment:
```suggestion
		batch    = u.conf.Batches[u.crtBatch-1]
		resultCh = make(chan error, len(batch))
```

Is this understanding correct?

Reply:
parallel always equals to length of batch, except the last batch which might be less than parallel